### PR TITLE
GODRIVER-2531 return canceled error message when canceled for waiting connection in wait queue

### PR DESCRIPTION
**Pull request by proost**
[GODRIVER-2531](https://jira.mongodb.org/browse/GODRIVER-2531)

when context is canceled waiting for connection, returns canceled error not timeout error.

duplicated codes are quite a lot. so how do you think refactor `WaitQueueTimeoutError` to `WaitQueueError` and And error message is different by error type?

### CODE REVIEW

**Review by qingyang-hu — CHANGES_REQUESTED** [review]
Hello @proost! We appreciate your PR. It's a reasonable enough addition. A more accurate error message in the case of Context cancellation could be helpful. However, I think adding a whole new error type is unnecessary because the `WaitQueueTimeoutError` already wraps the Context error, which will indicate it was a Context cancellation. Also, it is a breaking change for users who already uses the wrapped error to detect cancellation. Instead, we can modify the error message when the wrapped context error is `context.Canceled`.

**Comment by proost** [comment]
@qingyang-hu 
Thank you for your review. I agree we should avoid breaking change. so i change to use different error message when wrapped error is `context.Canceled`

File: x/mongo/driver/topology/pool_test.go [hunk 1]
```diff
@@ -645,6 +646,7 @@ func TestPool(t *testing.T) {
 			assert.IsTypef(t, WaitQueueTimeoutError{}, err, "expected a WaitQueueTimeoutError")
 			if err, ok := err.(WaitQueueTimeoutError); ok {
 				assert.Equalf(t, context.DeadlineExceeded, err.Unwrap(), "expected wrapped error to be a context.Timeout")
+				assert.Containsf(t, err.Error(), "timed out", "expected error to contain string")
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:649]:
```suggestion
				assert.Containsf(t, err.Error(), "timed out", `expected error message to contain "timed out"`)
```

File: x/mongo/driver/topology/pool_test.go [hunk 2]
```diff
@@ -773,6 +775,45 @@ func TestPool(t *testing.T) {
 			assert.Equalf(t, 2, p.totalConnectionCount(), "pool should have 2 total connection")
 			assert.Equalf(t, 0, p.availableConnectionCount(), "pool should have 0 idle connection")
 
+			p.close(context.Background())
+		})
+		t.Run("cancel wait queue", func(t *testing.T) {
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:780]:
```suggestion
		t.Run("canceled context in wait queue", func(t *testing.T) {
```
Could we use a more descriptive test name?

File: x/mongo/driver/topology/pool_test.go [hunk 3]
```diff
@@ -773,6 +775,45 @@ func TestPool(t *testing.T) {
...
+			})
+
+			p := newPool(poolConfig{
+				Address:     address.Address(addr.String()),
+				MaxPoolSize: 1,
+			})
+			err := p.ready()
+			noerr(t, err)
+
+			// check out first connection.
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:797]:
```suggestion
			// Check out first connection.
```

File: x/mongo/driver/topology/pool_test.go [hunk 4]
```diff
@@ -773,6 +775,45 @@ func TestPool(t *testing.T) {
...
+			_, err = p.checkOut(context.Background())
+			noerr(t, err)
+
+			// Set a short timeout and check out again.
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:801]:
```suggestion
			// Use a canceled context to check out another connection.
```
I don't think you need the goroutine to cancel the context, can we just do:
```
cancelCtx, cancel := context.WithCancel(context.Background())
cancel()
_, err = p.checkOut(cancelCtx)
```

File: x/mongo/driver/topology/pool_test.go [hunk 5]
```diff
@@ -773,6 +775,45 @@ func TestPool(t *testing.T) {
...
+			cancelCtx, cancel := context.WithCancel(context.Background())
+			go func() {
+				time.Sleep(100 * time.Millisecond)
+				cancel()
+			}()
+			_, err = p.checkOut(cancelCtx)
+			assert.NotNilf(t, err, "expected a WaitQueueCancel error")
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:808]:
```suggestion
			assert.NotNilf(t, err, "expected a non-nil error")
```

File: x/mongo/driver/topology/pool_test.go [hunk 6]
```diff
@@ -773,6 +775,45 @@ func TestPool(t *testing.T) {
...
+
+			// Assert that error received is WaitQueueTimeoutError with context canceled.
+			assert.IsTypef(t, WaitQueueTimeoutError{}, err, "expected a WaitQueueCancelError")
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:811]:
```suggestion
			assert.IsTypef(t, WaitQueueTimeoutError{}, err, "expected a WaitQueueTimeoutError")
```

File: x/mongo/driver/topology/pool_test.go [hunk 7]
```diff
@@ -773,6 +775,45 @@ func TestPool(t *testing.T) {
...
+			if err, ok := err.(WaitQueueTimeoutError); ok {
+				assert.Equalf(t, context.Canceled, err.Unwrap(), "expected wrapped error to be a context.Canceled")
+				assert.Containsf(t, err.Error(), "canceled", "expected error to contain string")
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:814]:
```suggestion
				assert.Containsf(t, err.Error(), "canceled", `expected error message to contain "canceled"`)
```

**Review by benjirewis — CHANGES_REQUESTED** [review]
Thanks so much for your contribution @proost 😄 . The code changes LGTM, I just had a few style comments on the new test and a potential simplification.

File: x/mongo/driver/topology/pool_test.go (hunk 1 above)
Comment (proost) [reply x/mongo/driver/topology/pool_test.go]: [ef458c5](https://github.com/mongodb/mongo-go-driver/pull/1058/commits/ef458c51dfccd3237a7d3454669196014aabeb1d)

File: x/mongo/driver/topology/pool_test.go (hunk 2 above)
Comment (proost) [reply x/mongo/driver/topology/pool_test.go]: [ef458c5](https://github.com/mongodb/mongo-go-driver/pull/1058/commits/ef458c51dfccd3237a7d3454669196014aabeb1d)

File: x/mongo/driver/topology/pool_test.go (hunk 3 above)
Comment (proost) [reply x/mongo/driver/topology/pool_test.go]: [ef458c5](https://github.com/mongodb/mongo-go-driver/pull/1058/commits/ef458c51dfccd3237a7d3454669196014aabeb1d)

File: x/mongo/driver/topology/pool_test.go (hunk 4 above)
Comment (proost) [reply x/mongo/driver/topology/pool_test.go]: [ef458c5](https://github.com/mongodb/mongo-go-driver/pull/1058/commits/ef458c51dfccd3237a7d3454669196014aabeb1d)

File: x/mongo/driver/topology/pool_test.go (hunk 6 above)
Comment (proost) [reply x/mongo/driver/topology/pool_test.go]: [ef458c5](https://github.com/mongodb/mongo-go-driver/pull/1058/commits/ef458c51dfccd3237a7d3454669196014aabeb1d)

File: x/mongo/driver/topology/pool_test.go (hunk 7 above)
Comment (proost) [reply x/mongo/driver/topology/pool_test.go]: [ef458c5](https://github.com/mongodb/mongo-go-driver/pull/1058/commits/ef458c51dfccd3237a7d3454669196014aabeb1d)

File: x/mongo/driver/topology/pool_test.go [hunk 8]
```diff
@@ -15,6 +15,7 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+
```
Comment (benjirewis) [inline x/mongo/driver/topology/pool_test.go:18]:
```suggestion
```

**Review by benjirewis — APPROVED** [review]
LGTM! Let's wait for linked patch to pass and for @qingyang-hu to review. Thanks again for your contribution, @proost 🧑‍🔧 .

File: x/mongo/driver/topology/pool_test.go (hunk 8 above)
Comment (proost) [reply x/mongo/driver/topology/pool_test.go]: [c474832](https://github.com/mongodb/mongo-go-driver/pull/1058/commits/c4748327cd33f4035beaaac427511bcbb2015a0a)

fix it. My IDE setting inserts new line. I wish there was a mention in `CONTRIBUTING.md` for like a this case.

Comment (benjirewis) [reply x/mongo/driver/topology/pool_test.go]: Ah I assume your IDE is set up to separate external and internal imports with a whitespace. We do recommend in `CONTRIBUTING.md` that contributors run `make fmt`, which should get rid of issues like this.

**Pull request merged**